### PR TITLE
Update micro-CDR API

### PR DESCRIPTION
**Pull request by TSC21**
This continues https://github.com/PX4/Firmware/pull/9645 and bumps up the micro-CDR submodule. At the same time, adapts the templates for the micro-RTPS agent and client code.

**Status: 12JUN**
It currently builds correctly but when running both agent and client apps, I am not able to get a Publisher and Subscriber match, so probably some detail missing atm.
To reproduce, on SITL, follow: https://dev.px4.io/en/middleware/micrortps.html. To start the agent: `./micrortps_agent -t UDP`. To start the client, on the px4 shell: `micrortps_client start -t UDP`.

### CODE REVIEW

File: msg/templates/uorb_microcdr/msg.cpp.template [hunk 1]
```diff
@@ -141,24 +140,25 @@ def add_code_to_deserialize():
...
 
-    resetStaticMicroCDRForSerialize(microCDRWriter);
+    reset_micro_buffer(microCDRWriter);
 
-    serializeUnsignedLong(input->timestamp, microCDRWriter);
+    serialize_uint64_t(microCDRWriter, input->timestamp);
 @add_code_to_serialize()
-
-    (*length) = microCDRWriter->m_microBuffer->m_serializedBuffer;
+    (*length) = (*microCDRWriter->final);
```
Comment (BorjaOuterelo) [inline msg/templates/uorb_microcdr/msg.cpp.template]: Assigning the final value to length is a mistake.
`microCDRWriter->final` points one position pass the last serialised element.
Dereferencing it does not gives you the length.
You need to get the distance between init and final pointers:
  `  (*length) = (microCDRWriter->final - microCDRWriter->init);`

**Comment by TSC21** [comment]
@dagar any suggestion on how to solve the endianness check problem on the `px4fmuv3_rtps` build? I tried to remove the endianness check as you can see, but that results on a build problem. With the endianness check, it does fails to check for big endian in the toolchain during the Cmake check.

**Comment by TSC21** [comment]
@BorjaOuterelo while compiling the target that builds the `micrortps_bridge` app (which includes micro-CDR), and during the endianness check, we have a resulting compiling error:
```bash
make nuttx_px4fmu-v3_rtps
-- PX4 VERSION: first_pos_control_flight-22839-g7d26cbb
-- CONFIG: nuttx_px4fmu-v3_rtps
-- Build Type: MinSizeRel
-- The ASM compiler identification is GNU
-- Found assembler: /home/tsc21/gcc-arm-none-eabi-5_4-2016q2/bin/arm-none-eabi-gcc
-- Found PythonInterp: /usr/bin/python (found version "2.7.12") 
-- Found PY_jinja2: /usr/local/lib/python2.7/dist-packages/jinja2  
-- C compiler: arm-none-eabi-gcc (GNU Tools for ARM Embedded Processors) 5.4.1 20160609 (release) [ARM/embedded-5-branch revision 237715]
-- C++ compiler: arm-none-eabi-g++ (GNU Tools for ARM Embedded Processors) 5.4.1 20160609 (release) [ARM/embedded-5-branch revision 237715]
-- PX4 ECL: Very lightweight Estimation & Control Library v0.9.0-558-g1cba257
-- Building and including px4io-v2
-- Using C++03
-- Release build type: MinSizeRel
-- Adding UAVCAN STM32 platform driver
-- Check if the system is big endian
-- Searching 16 bit integer
-- Looking for sys/types.h
CMake Error at /home/tsc21/PX4/Firmware_dev/src/pr-microcdr_update/cmake/toolchains/Toolchain-arm-none-eabi.cmake:83 (message):
  Processor not recognised in toolchain file
Call Stack (most recent call first):
  /home/tsc21/PX4/Firmware_dev/src/pr-microcdr_update/build/nuttx_px4fmu-v3_rtps/CMakeFiles/3.5.2/CMakeSystem.cmake:6 (include)
  CMakeLists.txt:3 (project)


CMake Error: Internal CMake error, TryCompile configure of cmake failed
-- Configuring incomplete, errors occurred!
See also "/home/tsc21/PX4/Firmware_dev/src/pr-microcdr_update/build/nuttx_px4fmu-v3_rtps/CMakeFiles/CMakeOutput.log".
make[1]: *** /home/tsc21/PX4/Firmware_dev/src/pr-microcdr_update/build/nuttx_px4fmu-v3_rtps: No such file or directory.  Stop.
Makefile:148: recipe for target 'nuttx_px4fmu-v3_rtps' failed
make: *** [nuttx_px4fmu-v3_rtps] Error 2
```
It complains about a not recognized processor on the toolchain immediately after the endianness check. Our available targets are in https://github.com/PX4/Firmware/blob/master/cmake/toolchains/Toolchain-arm-none-eabi.cmake#L75-L84. Can this be something missing on the `CMakeLists.txt` for micro-CDR or is this PX4 build structure related? Do you have a suggestion on how to solve this?

**Comment by thomasgubler** [comment]
Talked to @dagar 
@dagar thinks that the issue might be on the px4 build system side. He has a WIP fix ready at https://github.com/PX4/Firmware/pull/9808 . However that one does not pass the NuttX CI tests. He will try to look into if he can fix that quickly so that we can try this PR #9561 with #9808

**Comment by TSC21** [comment]
@thomasgubler thanks for the heads-up!

**Comment by BorjaOuterelo** [comment]
Hi @TSC21 

Micro CDR cmake has a macro that controls if endianness is checked or not.

In the version you using in master this option was configured as OFF by default
https://github.com/eProsima/micro-CDR/blob/50776f4ab7895e47af535a6d6e7ec697ddf443f3/CMakeLists.txt#L127

But in Micro CDR master it was changed to be ON by default.
https://github.com/eProsima/micro-CDR/blob/04623b535d0e774ffb4b42349a42a6bab0302ecf/CMakeLists.txt#L104

It is more evident in the diff: https://github.com/eProsima/micro-CDR/compare/50776f4ab7895e47af535a6d6e7ec697ddf443f3...0061c9af1f17e560bee604dd882772913bf90073

This change may be triggering the endianness check and thus probably causing some issues.

**Comment by BorjaOuterelo** [comment]
@TSC21 you are right, I solved those Cmake issues on my side, and now I have some compile errors due to warnings on Micro CDR.
I am currently fixing those Micro CDR warnings.
I will keep you updated on my progress.

**Comment by TSC21** [comment]
@BorjaOuterelo thanks. I guess then this is not entirely on the PX4 side, cause it does actually launch build errors on the micro-CDR build process when the endianness check is off.

**Comment by BorjaOuterelo** [comment]
I have made all the changes on Micro CDR to fix errors while compiling PX4.
@TSC21 can you try with the branch I pushed? https://github.com/eProsima/micro-CDR/tree/px4
We will merge this branch into master before tagging a version.

**Comment by TSC21** [comment]
> I have made all the changes on Micro CDR to fix PX4 compiling errors.
@TSC21 can you try with the branch I pushed? https://github.com/eProsima/micro-CDR/tree/px4
We will merge this branch into master before tagging a version.

That did the trick! Thanks. Though the temporary solution is the set the endianness check to off, that is not the solution we want further. Though, while there aren't developments in #9808, and while it's not merged, I think this PR can go in.

**Comment by TSC21** [comment]
@dagar good to go?

**Comment by dagar** [comment]
I'm not setup to test this, but it looks good. Let's talk about getting something basic in CI later.

**Comment by TSC21** [comment]
A little detail: we are using the "px4" branch of https://github.com/eProsima/micro-CDR in this. @BorjaOuterelo please let us know when that gets to master and tagged so the submodule tag can be updated accordingly. Thanks!

**Comment by BorjaOuterelo** [comment]
Hi, @TSC21 micro-CDR master branch includes the changes made on the "px4" branch, and we have created a tag.
https://github.com/eProsima/micro-CDR/tree/v1.0.0beta2

Feel free to update your submodule accordingly; It also includes some other changes, the ones possibly affecting you are the change of the return type of micro_buffer functions from uint32_t to size_t:

```
microcdr_DllAPI size_t micro_buffer_size(const MicroBuffer* buffer);
microcdr_DllAPI size_t micro_buffer_length(const MicroBuffer* buffer);
microcdr_DllAPI size_t micro_buffer_remaining(const MicroBuffer* buffer);
```

Apart from the return type, We also changed the parameter of those functions to be const.

Best.

**Comment by TSC21** [comment]
@BorjaOuterelo thanks for the update. I will update the code accordingly as soon as possible.

**Pull request merged**